Add the capability to analyze wave height and 10-m wind gust into HRRR-3DRTMA package

GangZhao-NOAA opened the pull request:
***HRRR-3DRTMA package***
1. GSI source code would be git-cloned from github repository;
2. Add steps in workflow (ex-scripts, j-job scripts, xml files;
3. some updated fix files (anavonfo,convinfo under fix/gsi) and parm file (gsiparm.anl.sh under parm/gsi) are NOT included in this pacakge.
4.  For information on the Motivations and Methods, see Issue #32 **

***Related Issues.***
- This Pull request is to resolve issue #32 
Resolves #32  

***Test***
- Workflow has been tested for CONUS domain and ALaska domain on wcoss2 Dogwood in real-time parallel run.

GangZhao-NOAA committed:
Add the capability to analyze wave height and 10-m wind gust into
HRRR-3DRTMA package.
1. GSI source code would be git-cloned from github repository;
2. Add steps in workflow (ex-scripts, j-job scripts, xml files;
3. some updated fix files (anavonfo,convinfo under fix/gsi) and parm file
   (gsiparm.anl.sh under parm/gsi) are NOT included in this pacakge.

AnnetteGibbs-NOAA reviewed (approved):
Looks good Gang!

MatthewMorris-NOAA commented on scripts/rtma3d/exrtma3d_gsianl.ksh:
```diff
@@ -380,8 +464,28 @@ if [ "${envir}" == "lsf" ] || [ "${envir}" == "pbspro" ]; then #WCOSS
 fi
+
+#====  set GSI namelist options for analysis of HOWV and GUST  ====#
+  oerr_gust="-1.0"                #Obs Err of gust (if<0, use preset value 1.0 defined in read_prepbufr.f90)
```
Should this value be negative?

GangZhao-NOAA replied:
@MatthewMorris-NOAA,
Hi Matthew,

if set it be a negative value, that means GSI would use its default value. I understand your concern that a negative value here is really confusing. Maybe I'd better to remove it in the namelist, so GSI still runs with the default value, and users would be not be confused with the negative value.

thanks,
-Gang

MatthewMorris-NOAA commented on workflow/rtma3d_pbspro_conus.xml:
```diff
@@ -18,21 +18,25 @@
 <!-- Variables Defined by absolute paths -->
 
-<!ENTITY ptmp_base	"/lfs/h2/emc/ptmp/&USER;/3DRTMA_HRRR_CONUS_NEW_WRKDIR">
+<!ENTITY ptmp_base	"/lfs/h2/emc/ptmp/&USER;/3DRTMA_HRRR_CONUS&ExpName;_WRKDIR">
```
Suggest putting an underscore before ExpName, i.e., 3DRTMA_HRRR_CONUS_&ExpName;_WRKDIR
Same for NWROOT and LOG_DIR below.

GangZhao-NOAA replied:
@MatthewMorris-NOAA 
Hi Matthew,

I agree with you. Actually in the beginning I put the "_" sign before ExpName. Then I found that if I did not define ExpName, then there would be an extra underscore sign "_" in all the directory name. 

So using "Expname" is kind of clumsy and is easy to cause problem, so I am going to remove this "ExpName" in the xml file. Then users could define the "ptmp_base", "NWROOT" and "LOG_DIR" for their own experimental systems.

@AnnetteGibbs-NOAA @EdwardColon-NOAA, what do you think about this?

Thanks,
-Gang

AnnetteGibbs-NOAA replied:
@GangZhao-NOAA Sounds good to me!  That's what I do - I just define the directories for "ptmp_base", "NWROOT" and "LOG_DIR".

MatthewMorris-NOAA commented on workflow/rtma3d_pbspro_conus.xml:
```diff
@@ -216,22 +223,25 @@
-<!ENTITY PREPFGS_RESOURCES '<walltime>00:10:00</walltime>'>
-<!ENTITY PREPFGS_RESERVATION '<nodes>1:ppn=1:vntype=cray_compute:mem=4000MB</nodes><native>-l place=scatter</native><queue>&QUEUE;</queue><account>&ACCOUNT;</account>'>
+<!ENTITY PREPFGS_RESOURCES '<walltime>00:15:00</walltime>'>
+<!ENTITY PREPFGS_RESERVATION '<nodes>1:ppn=1:vntype=cray_compute:mem=50GB</nodes><native>-l place=scatter</native><queue>&QUEUE;</queue><account>&ACCOUNT;</account>'>
```
Why do we need this increase in memory?

GangZhao-NOAA replied:
@MatthewMorris-NOAA 
Hi Matthew,

I add the step to dump out howv and gust from grib2 to binary (by using wgrib2) and then append the binary data to netcdf firstguess file (by using ncks). the ncks step would generate a temporary 16GB file which requires large memory. So I have to increase the memory usage for this step.

Thank you!

-Gang

MatthewMorris-NOAA replied:
Sounds good!

GangZhao-NOAA replied:
@MatthewMorris-NOAA 
I tested with mem=20GB for CONUS domain on wcoss2. Code does not crash, but ran with warning message "group memory limit exceeded". So I switch back to 50GB.
Thank you!
-Gang

EdwardColon-NOAA commented on workflow/rtma3d_pbspro_conus.xml:
```diff
@@ -904,7 +922,7 @@
 
   <cycledef group="rap">0 0,12 * 01-12 2022-2025 *</cycledef>
 
-  <task name="&RUN;_obsprep_lghtn_task_@Y@m@d@H@M" cycledefs="hourly" maxtries="&maxtries;">
```
Hi Gang. I think that the date/time string should be kept with each task name since multiple cycles can and do run at the same time. It would be difficult to distinguish which cycle is running for a given workflow step and if there is a problem, it may be hard to distinguish in which cycle it occurred.

AnnetteGibbs-NOAA replied:
@EdwardColon-NOAA I agree with removing the date/time string from the task name.  Whenever I rocotoboot, I have to do the following:

rocotoboot -w rtma3d_pbspro_conus.xml -d rtma3d_pbspro_conus.db -c 202503051100 -t rtma3d_prdgen_task_@Y@m@d@H@M 

as "@Y@m@d@H@M" is part of the task name (and it does not fill in the date).  If you look in Gang's log directories, it does have the date in the logfile name.  So I think it's unneccesary to include "@Y@m@d@H@M" in the task name.

GangZhao-NOAA replied:
@EdwardColon-NOAA ,
Hi Ed,

As @AnnetteGibbs-NOAA mentioned in her comments above, the time stamp string "_@Y@m@d@H@M" actually does NOT work. It in fact did not show the time  string in the task name (because the "<cyclestr> is NOT added in the task name definition yet). But it is not necessary, since the time stamp already show for each task when rocoto lists each task. 

And removing this time stamp in task name has no influence to the log file name. The log file still has the time stamp (since the jobname is defined with timestamp as follows):

`<jobname><cyclestr>&RUN;_obsprep_cloud_job_@Y@m@d@H@M</cyclestr></jobname>`

So I believe it is good to remove the time stamp in the task name definition. 

Thanks,
-Gang

EdwardColon-NOAA commented on scripts/rtma3d/exrtma3d_anldump.ksh:
This new ex-script looks good.

EdwardColon-NOAA commented on scripts/exrtma3d_GribMerge_urma.sh:
The GribMerge script is well documented. No issue.

EdwardColon-NOAA commented on jobs/JRTMA3D_ANLDUMP:
Looks good.

EdwardColon-NOAA reviewed (commented):
Hi Gang. I only see an issue in the xml file.

GangZhao-NOAA committed:
1. Following reviewers' suggestions, made some modifications;
2. adding howv & gust support for Alaksa;
3. adding scripts/exrtma3d_filter_land.sh to use the sea-land mask when generating firstguess of howv;

GangZhao-NOAA commented:
@AnnetteGibbs-NOAA @MatthewMorris-NOAA @EdwardColon-NOAA @ManuelPondeca-NOAA 
Hi all,
Following your suggestions, I made some modifications, and added the howv/gust support to Alaska run. I tested this workflow for both CONUS and Alaska on WCOSS2. It works good. So I made a new commit and updated this Pull Request #31 . 
Please give it another round of review.
Thank you!
-Gang

AnnetteGibbs-NOAA reviewed (approved):
Everything looks good Gang.  Thank you for making the changes and testing the Alaska grid.

MatthewMorris-NOAA commented on scripts/akrtma3d/exakrtma3d_gsianl.ksh:
```diff
@@ -222,6 +283,27 @@ OZINFO=${FIXgsi}/global_ozinfo.txt
+ANAVINFO_GUST_FN=anavinfo_arw_netcdf_gust
+CONVINFO_HOWVGUST_FN=3drtma_convinfo_v0.6.5_updated_howvgust
+CONVINFO_HOWV_FN=3drtma_convinfo_v0.6.5_updated_howvgust
+CONVINFO_GUST_FN=3drtma_convinfo_v0.6.5_updated_howvgust
```
@GangZhao-NOAA What is the reason for having separate convinfo files for howvgust, howv, and gust?

GangZhao-NOAA replied:
@MatthewMorris-NOAA 
Hi Matthew,
I agree with you that we should keep it simple. I set up 3 different anavinfo files just for testing run if howv, or gust is missing in the run, then only use the matching anavinfo file which would not lead to analysis crash. In final version for operational run, we may keep only one anavinfo for 3DRTMA.
Thank you!
-Gang

EdwardColon-NOAA commented:
Hi Gang,

I approve of the most recent updates.

GangZhao-NOAA commented:
@EdwardColon-NOAA @AnnetteGibbs-NOAA @MatthewMorris-NOAA @ManuelPondeca-NOAA 
Hi Ed,

Honestly I have not done with the step "Merge pull request" before. I do not know if I have the privilege to merge PR. Should I just click that green button "Merge pull request"? Or it should be you who is the repo manager to do this?

Thank you!

-Gang
